File: front-end/Mobile/src/models/attendance/SupportingDocument.js
import BaseModel from '../BaseModel';

export default class SupportingDocument extends BaseModel {
  constructor(data = {}) {
    super(data);
    this.supportingDocumentId = data.supporting_document_id || null;
    this.justificationId = data.justification_id || null;
    this.fileName = data.file_name || '';
    this.storageUri = data.storage_uri || '';
    this.mimeType = data.mime_type || '';
    this.sizeBytes = data.size_bytes || 0;
  }

  get sizeFormatted() {
    if (this.sizeBytes < 1024) return `${this.sizeBytes} B`;
    if (this.sizeBytes < 1024 * 1024) return `${(this.sizeBytes / 1024).toFixed(1)} KB`;
    return `${(this.sizeBytes / (1024 * 1024)).toFixed(1)} MB`;
  }

  static fromApi(data) {
    if (!data) return null;
    return new SupportingDocument(data);
  }

  toApi() {
    return {
      ...super.toApi(),
      supporting_document_id: this.supportingDocumentId,
      justification_id: this.justificationId,
      file_name: this.fileName,
      storage_uri: this.storageUri,
      mime_type: this.mimeType,
      size_bytes: this.sizeBytes,
    };
  }
}

File: back-end/05-ms-attendance/src/main/java/com/faceattend_edu/attendance_service/domain/model/SupportingDocument.java
package com.faceattend_edu.attendance_service.domain.model;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

@Getter @Setter @AllArgsConstructor @NoArgsConstructor
public class SupportingDocument {
    private Long supportingDocumentId;
    private Long justificationId;
    private String fileName;
    private String storageUri;
    private String mimeType;
    private Long sizeBytes;
    private Instant createdAt;
    private Instant updatedAt;
    private Instant deletedAt;
    private UUID createdBy;
    private UUID updatedBy;
    private UUID deletedBy;
    private long rowVersion;

    public void validate() {
        if (justificationId == null) throw new IllegalArgumentException("justificationId is required");
        if (isBlank(fileName)) throw new IllegalArgumentException("fileName is required");
        if (fileName.length() > 255) throw new IllegalArgumentException("fileName max 255");
        if (isBlank(storageUri)) throw new IllegalArgumentException("storageUri is required");
        if (storageUri.length() > 500) throw new IllegalArgumentException("storageUri max 500");
        if (isBlank(mimeType)) throw new IllegalArgumentException("mimeType is required");
        if (sizeBytes == null || sizeBytes <= 0) throw new IllegalArgumentException("sizeBytes must be > 0");
    }
    public void touchCreated(){ if(createdAt==null) createdAt=Instant.now(); if(rowVersion==0) rowVersion=1L; }
    public void touchUpdated(){ updatedAt=Instant.now(); }
    private static boolean isBlank(String s){ return s==null || s.trim().isEmpty(); }
    @Override public boolean equals(Object o){ if(this==o) return true; if(!(o instanceof SupportingDocument)) return false; SupportingDocument that=(SupportingDocument)o; return Objects.equals(supportingDocumentId, that.supportingDocumentId); }
    @Override public int hashCode(){ return Objects.hash(supportingDocumentId); }
}
